### PR TITLE
policy: refactor `main.rs` into `lib.rs`

### PR DESCRIPTION
**Pull request by hawkw**
Currently, the `linkerd-policy-controller` crate contains a `lib.rs` and a `main.rs`. The `lib.rs` defines some of the more generic interfaces into the policy-controller's application logic, but a substantial amount of process startup (e.g. handling CLI arguments, configuring clients and servers, spawning tasks, etc) is performed in the `main` function in `main.rs`.

This factoring feels somewhat awkward. If it's not really possible to run a policy-controller process without code that only exists in `main.rs`, why provide a `lib.rs` interface at all? We may as well move all of the code necessary to run a policy-controller into the library crate, and have a very simple `main.rs` that simply calls into the library's top-level run function. In the future, this could potentially enable end-to-end testing by running multiple policy-controller binaries in a single process, or allow downstream code to run the policy-controller as part of other software that interfaces with Linkerd.

This branch refactors the policy-controller to move all the application logic currently defined in `main.rs` into the `lib.rs` as a public `run` method on the CLI arguments type, and a free function `run()` that parses CLI args and then runs the controller.

The `main.rs` file now need only contain a three-line `main` function that just does this:

```rust
fn main() -> anyhow::Result<()> {
    linkerd_policy_controller::run()
}
```

(along with some crate-level attributes and the `static` for using `jemalloc` as the system allocator)

### CODE REVIEW

**Review by adleong — APPROVED** [review]
TIOLI: lib.rs is getting a bit long here.  The Args code might make sense to split into an args.rs?

File: policy-controller/src/lib.rs [hunk 1]
```diff
@@ -3,25 +3,298 @@
...
+use tracing::{info, info_span, instrument, Instrument};
+
+const DETECT_TIMEOUT: Duration = Duration::from_secs(10);
+const LEASE_DURATION: Duration = Duration::from_secs(30);
+const LEASE_NAME: &str = "policy-controller-write";
+const RENEW_GRACE_PERIOD: Duration = Duration::from_secs(1);
+
+#[derive(Debug, Parser)]
+#[clap(name = "policy", about = "Linkerd 2 policy controller")]
+pub struct Args {
```
Comment (olix0r) [inline policy-controller/src/lib.rs:39]: It's probably only useful for this to be _pub_ if the individual arguments/subtypes are as well.

**Comment by olix0r** [comment]
I think we ultimately would be better off actually having some logic in the main (akin to what we have in the proxy -- things that are truly about setting up the runtime, while pushing the actual business logic into the library. I.e. i'm not convinced that moving Args into the library is as useful as moving the indexers. I think we could also split up the indexer code so it's not a big monolithic thing (AIUI the inbound and outbound indexers are basically completely separate currently).

**Comment by hawkw** [comment]
@olix0r my thinking was that we would start by, essentially, moving everything into `lib.rs` behind One Big `run()` function, and then start to break out other public APIs, such as running the indexers, as needed; the goal here is primarily that it should be possible to depend on "some way to run a policy controller" as a library, and then extend the interface of that library to allow more granular configuration, as needed.

Happy to take a different approach, though, if that's not what you think is best!

**Pull request closed without merging**